### PR TITLE
Sync play status with HBO GO (EU handler) (#88)

### PR DESCRIPTION
**Pull request by arvvoid**
* Send and get play status from Hbo Go (EU Handler)

## Please check that your pull request pass this checks:

-  [x] My changes respect the [Kodi add-on development rules](https://kodi.wiki/view/Add-on_rules)
-  [x] My code is both Python 2 and 3 compatible
-  [x] My code follows the code style of this project
-  [x] I made sure there wasn't another Pull Request opened for the same update/change
-  [x] I have successfully tested my changes locally

## Types of changes

-  [ ] Bug fix (non-breaking change which fixes an issue)
-  [x] New feature (non-breaking change which adds functionality)
-  [ ] Breaking change (fix or feature that would cause existing functionality to change)
-  [ ] Adding a new region/hbo go version

## Description:

Report and get play status from Hbo go (EU Handler)

### CODE REVIEW

**Comment by arvvoid** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- hbogolib/handlereu.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315&bid=15689276)

File: hbogolib/handlereu.py [hunk 1]
```diff
@@ -1234,6 +1305,15 @@ def addLink(self, title, mode):
         liz.addStreamInfo('video', {'aspect': 1.78, 'codec': 'h264'})
         liz.addStreamInfo('audio', {'codec': 'aac', 'channels': 2})
         liz.setProperty("IsPlayable", "true")
+        if hbogo_position > -1:
+            self.log("Found elapsed time on Hbo go for " + cid + " External ID: " + externalid + " Elapsed: " + str(hbogo_position) + " of " + str(title[
+                                                                                                                                                   'Duration']))
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [line too long (160 > 159 characters)](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

**Comment by arvvoid** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- hbogolib/base.py  2
- hbogolib/handlereu.py  5
         

Clones added
============
- hbogolib/handlereu.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315&bid=15689276)

File: hbogolib/handlereu.py [hunk 2]
```diff
@@ -1146,6 +1191,22 @@ def genContextMenu(self, content_id, media_id):
         remove_mylist = (py2_encode(self.language(30720)), runplugin %
                          (self.base_url, remove_mylist_query))
 
+        watched_query = urlencode({
+            'url': 'WATCHED',
+            'mode': HbogoConstants.ACTION_MARK_WATCHED,
+            'cid': content_id,
+        })
+        mark_watched = (py2_encode(self.language(30803)), runplugin %
+                         (self.base_url, watched_query))
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Wrong continued indentation (remove 1 space).](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

File: hbogolib/handlereu.py (hunk 1 above)
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Wrong hanging indentation (remove 131 spaces).](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

File: hbogolib/handlereu.py [hunk 3]
```diff
@@ -1146,6 +1191,22 @@ def genContextMenu(self, content_id, media_id):
...
+
+        unwatched_query = urlencode({
+            'url': 'UNWATCHED',
+            'mode': HbogoConstants.ACTION_MARK_UNWATCHED,
+            'cid': content_id,
+        })
+        mark_unwatched = (py2_encode(self.language(30804)), runplugin %
+                         (self.base_url, unwatched_query))
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Wrong continued indentation (add 1 space).](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

File: hbogolib/handlereu.py [hunk 4]
```diff
@@ -1156,21 +1217,31 @@ def genContextMenu(self, content_id, media_id):
...
-            return list(votes) + [remove_mylist]
-
-        return [add_mylist] + list(votes)
+                      'url': 'VOTE',
+                      'mode': HbogoConstants.ACTION_VOTE,
+                      'vote': item['vote'],
+                      'cid': content_id,
+                  }))) for item in votes_configs]
+
+        if mode == HbogoConstants.CONTEXT_MODE_MOVIE:
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

File: hbogolib/handlereu.py [hunk 5]
```diff
@@ -1156,21 +1217,31 @@ def genContextMenu(self, content_id, media_id):
...
+            if self.cur_loc == self.LB_MYPLAYLIST:
+                return [mark_watched, mark_unwatched] + list(votes) + [remove_mylist]
+            else:
+                return [mark_watched, mark_unwatched, add_mylist] + list(votes)
+        elif mode == HbogoConstants.CONTEXT_MODE_EPISODE:
+            return [mark_watched, mark_unwatched]
+        else:
+            if self.cur_loc == self.LB_MYPLAYLIST:
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

**Comment by arvvoid** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- hbogolib/base.py  2
- hbogolib/handlereu.py  5
         

Clones added
============
- hbogolib/handlereu.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315&bid=15689276)

File: hbogolib/handlereu.py [hunk 6]
```diff
@@ -1156,21 +1216,40 @@ def genContextMenu(self, content_id, media_id):
...
+                    return list(votes) + [remove_mylist]
+                else:
+                    return [add_mylist] + list(votes)
+        elif mode == HbogoConstants.CONTEXT_MODE_EPISODE:
+            if self.addon.getSetting('send_elapsed') == 'true':
+                return [mark_watched, mark_unwatched]
+            else:
+                return []
+        else:
+            if self.cur_loc == self.LB_MYPLAYLIST:
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

File: hbogolib/handlereu.py [hunk 7]
```diff
@@ -1156,21 +1216,40 @@ def genContextMenu(self, content_id, media_id):
...
+                    return [mark_watched, mark_unwatched] + list(votes) + [remove_mylist]
+                else:
+                    return [mark_watched, mark_unwatched, add_mylist] + list(votes)
+            else:
+                if self.cur_loc == self.LB_MYPLAYLIST:
+                    return list(votes) + [remove_mylist]
+                else:
+                    return [add_mylist] + list(votes)
+        elif mode == HbogoConstants.CONTEXT_MODE_EPISODE:
+            if self.addon.getSetting('send_elapsed') == 'true':
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

File: hbogolib/handlereu.py [hunk 8]
```diff
@@ -1156,21 +1216,40 @@ def genContextMenu(self, content_id, media_id):
...
-            return list(votes) + [remove_mylist]
-
-        return [add_mylist] + list(votes)
+                      'url': 'VOTE',
+                      'mode': HbogoConstants.ACTION_VOTE,
+                      'vote': item['vote'],
+                      'cid': content_id,
+                  }))) for item in votes_configs]
+
+        if mode == HbogoConstants.CONTEXT_MODE_MOVIE:
```
Comment (arvvoid) [inline hbogolib/handlereu.py]: Codacy Issue found: [Unnecessary "else" after "return"](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315)

**Comment by arvvoid** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- hbogolib/base.py  2
- hbogolib/handlereu.py  5
         

Clones added
============
- hbogolib/handlereu.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315&bid=15689276)

**Comment by arvvoid** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- hbogolib/base.py  2
- hbogolib/handlereu.py  5
         

Clones added
============
- hbogolib/handlereu.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315&bid=15689276)

**Comment by arvvoid** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- hbogolib/handlereu.py  5
- hbogolib/base.py  2
         

Clones added
============
- hbogolib/handlereu.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/arvvoid/plugin.video.hbogoeu/pullRequest?prid=4693315&bid=15689276)

**Comment by sile70000** [comment]
@arvvoid 
Wow! You're so fast.
Sorry I didn't have time to test much.
Still, I've tested this morning and the track_elapsed function still does not behave itself :).
So I've change it a bit and it looks like now is working like  I've wanted from the start.
Can you please take a look and share your thoughts.
Thanks.
[track_elapsed.txt](https://github.com/arvvoid/plugin.video.hbogoeu/files/3976749/track_elapsed.txt)

**Comment by arvvoid** [comment]
Hello, tnx for testing in all my quick tests was working fine but I didn't test heavily with playlist and slower devices. Like this should be fail-proof.
I disagree with eliminating the real-time update from the loop, I will keep that, that's the way the official client behaves. This way if the internet connection stop in middle or a crash happens in the middle the play status will still be saved. Since it updates only on % changes are still safe to use even if we loop more often.
The loop already as it was updated only on 1% changes not each loop so your avoiding to update if less then 11-sec playing will still work unless the video is super short.

**Comment by arvvoid** [comment]
I looked at it now you are right, the official client updates only on pause or stop. I will remove the real-time update or make it fire less often.

**Comment by arvvoid** [comment]
@sile70000 Made the changes, new version is in master.

**Pull request merged**